File: src/test/java/com/example/pages/CartPage.java
package com.example.pages;

import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;

/**
 * Page Object for cart page (/view_cart).
 */
public class CartPage extends BasePage {

    // Main cart page container
    private final By cartPageHeading = By.cssSelector("#cart_items, .cart_info, .table.table-condensed");
    // First product row in cart
    private final By firstCartProductRow = By.cssSelector("tr[id^='product-']");
    // Proceed To Checkout button
    private final By proceedToCheckoutButton = By.cssSelector("a[href='/checkout'], a.check_out");
    // Register / Login link in checkout modal (appears when cart has items, user not logged in)
    private final By registerLoginInModal = By.cssSelector(".modal-content a[href='/login']");
    // Quantity cell – may be button/span/text in different implementations
    private final By quantityCellInsideFirstRow = By.cssSelector(
            "tr[id^='product-'] td.cart_quantity, " +
            "tr[id^='product-'] td.cart_quantity button, " +
            "tr[id^='product-'] td.cart_quantity span"
    );

    public CartPage(WebDriver driver) {
        super(driver);
    }

    /** Waits for product to appear in cart (gives time for session update). */
    public void waitForProductToAppear() {
        wait.until(ExpectedConditions.presenceOfElementLocated(firstCartProductRow));
    }

    public boolean isCartPageVisible() {
        return isElementPresent(cartPageHeading);
    }

    public boolean isAnyProductVisible() {
        return isElementPresent(firstCartProductRow);
    }

    @Step("Click Proceed To Checkout")
    public void clickProceedToCheckout() {
        click(proceedToCheckoutButton);
    }

    @Step("Click Register / Login in checkout modal")
    public void clickRegisterLoginInCheckoutModal() {
        click(registerLoginInModal);
    }

    public int getFirstProductQuantity() {
        String text = readText(quantityCellInsideFirstRow).trim();
        // Text may contain other characters – extract digits only
        String digits = text.replaceAll("[^0-9]", "");
        return digits.isEmpty() ? 0 : Integer.parseInt(digits);
    }
}

File: pages/cart_page.py
"""Cart page – proceed to checkout, quantity."""
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.support import expected_conditions as EC

from pages.base_page import BasePage


class CartPage(BasePage):
    cart_page_heading = (By.CSS_SELECTOR, "#cart_items, .cart_info, .table.table-condensed")
    first_cart_product_row = (By.CSS_SELECTOR, "tr[id^='product-']")
    proceed_to_checkout_button = (By.CSS_SELECTOR, "a[href='/checkout'], a.check_out")
    register_login_in_modal = (By.CSS_SELECTOR, ".modal-content a[href='/login']")
    quantity_cell_inside_first_row = (
        By.CSS_SELECTOR,
        "tr[id^='product-'] td.cart_quantity, "
        "tr[id^='product-'] td.cart_quantity button, "
        "tr[id^='product-'] td.cart_quantity span"
    )

    def __init__(self, driver: WebDriver):
        super().__init__(driver)

    def wait_for_product_to_appear(self) -> None:
        self.wait.until(EC.presence_of_element_located(self.first_cart_product_row))

    def is_cart_page_visible(self) -> bool:
        return self.is_element_present(self.cart_page_heading)

    def is_any_product_visible(self) -> bool:
        return self.is_element_present(self.first_cart_product_row)

    def click_proceed_to_checkout(self) -> None:
        self.click(self.proceed_to_checkout_button)

    def click_register_login_in_checkout_modal(self) -> None:
        self.click(self.register_login_in_modal)

    def get_first_product_quantity(self) -> int:
        import re
        text = self.read_text(self.quantity_cell_inside_first_row).strip()
        digits = re.sub(r"[^0-9]", "", text)
        return int(digits) if digits else 0
